src/System.Private.CoreLib/shared/System/Array.cs

```diff
@@ -1464,127 +1660,85 @@ public static void Sort(Array keys, Array? items, int index, int length, ICompar
             if (keys.Length - (index - keysLowerBound) < length || (items != null && (index - keysLowerBound) > items.Length - length))
                 ThrowHelper.ThrowArgumentException(ExceptionResource.Argument_InvalidOffLen);
 
-            if (length > 1)
+            if (length <= 1)
+                return;
+
+            comparer ??= Comparer.Default;
+
+            if (keys is object[] objKeys)
             {
-                comparer ??= Comparer.Default;
+                object[]? objItems = items as object[];
+                if (items == null || objItems != null)
+                {
+                    new SorterObjectArray(objKeys, objItems, comparer).Sort(index, length);
+                    return;
+                }
+            }
 
-                if (comparer == Comparer.Default)
+            if (comparer == Comparer.Default)
+            {
+                switch (Type.GetTypeCode(keys.GetType().GetElementType()))
                 {
-                    switch (Type.GetTypeCode(keys.GetType().GetElementType()))
-                    {
-                        case TypeCode.Boolean:
-                            if (TryGenericSort(keys as bool[], items, index, length)) return;
-                            break;
-                        case TypeCode.Byte:
-                            if (TryGenericSort(keys as byte[], items, index, length)) return;
-                            break;
-                        case TypeCode.Char:
-                            if (TryGenericSort(keys as char[], items, index, length)) return;
-                            break;
-                        case TypeCode.Double:
-                            if (TryGenericSort(keys as double[], items, index, length)) return;
-                            break;
-                        case TypeCode.Int16:
-                            if (TryGenericSort(keys as short[], items, index, length)) return;
-                            break;
-                        case TypeCode.Int32:
-                            if (TryGenericSort(keys as int[], items, index, length)) return;
-                            break;
-                        case TypeCode.Int64:
-                            if (TryGenericSort(keys as long[], items, index, length)) return;
-                            break;
-                        case TypeCode.SByte:
-                            if (TryGenericSort(keys as sbyte[], items, index, length)) return;
-                            break;
-                        case TypeCode.Single:
-                            if (TryGenericSort(keys as float[], items, index, length)) return;
-                            break;
-                        case TypeCode.String:
```

Comment:
Removed String. It was not optimized in the original FCall either and it does not have that bad performance either.